src/moscot/backends/ott/_utils.py

```diff
@@ -38,9 +69,251 @@ def _compute_sinkhorn_divergence(
     return float(output.divergence)
 
 
+class RunningAverageMeter:
+    """Computes and stores the average value."""
+
+    def __init__(self) -> None:
+        self.reset()
+
+    def reset(self) -> None:
+        """Reset the meter."""
+        self.avg: float = 0.0
+        self.sum: float = 0.0
+        self.count: float = 0.0
+
+    def update(self, val: float, n: int = 1) -> None:
+        """Update the meter."""
+        self.sum += val * n
+        self.count += n
+        self.avg = self.sum / self.count
+
+
+@partial(jax.jit, static_argnames=["k"])
+def get_nearest_neighbors(
+    input_batch: jnp.ndarray, target: jnp.ndarray, k: int = 30  # type: ignore[name-defined]
+) -> Tuple[jnp.ndarray, jnp.ndarray]:  # type: ignore[name-defined]
+    """Get the k nearest neighbors of the input batch in the target."""
+    if target.shape[0] < k:
+        raise ValueError(f"k is {k}, but must be smaller or equal than {target.shape[0]}.")
+    pairwise_euclidean_distances = jnp.sqrt(jnp.sum((input_batch - target) ** 2, axis=-1))
+    negative_distances, indices = jax.lax.top_k(-1 * pairwise_euclidean_distances, k=k)
+    return -1 * negative_distances, indices
+
+
 def _filter_kwargs(*funcs: Callable[..., Any], **kwargs: Any) -> Dict[str, Any]:
     res = {}
     for func in funcs:
         params = inspect.signature(func).parameters
         res.update({k: v for k, v in kwargs.items() if k in params})
     return res
+
+
+@jtu.register_pytree_node_class
+class ConditionalDualPotentials:
+    r"""The conditional Kantorovich dual potential functions as introduced in :cite:`bunne2022supervised`.
+
+    :math:`f` and :math:`g` are a pair of functions, candidates for the dual
+    OT Kantorovich problem, supposedly optimal for a given pair of measures.
+
+    Parameters
+    ----------
+    f
+        The first conditional dual potential function.
+    g
+        The second conditional dual potential function.
+    cost_fn
+        The cost function used to solve the OT problem.
+    corr
+        Whether the duals solve the problem in distance form, or correlation
+        form (as used for instance for ICNNs, see, e.g., top right of p.3 in
+        :cite:`makkuva:20`)
+    """
+
+    def __init__(self, state_f: TrainState, state_g: TrainState):
```

Comment:
Consider subclassing the `DualPotentials` directly and adding the condition here to init + use `f/g` local function as defined, e.g., in `to_dual_potentials`.

Reply:
will adapt this once it's refactored in OTT-JAX